include/openrave/openravecontainer.h

```diff
@@ -0,0 +1,265 @@
+                _beginValidElementsIndex = 0;
+                _endValidElementsIndex = 0;
+            }
+            else {
+                // update begin and end index
+                if (index == _beginValidElementsIndex) {
+                    if (_numValidElements == 1) {
+                        _beginValidElementsIndex = _endValidElementsIndex - 1;
+                    }
+                    else {
+                        for (size_t newBeginIndex = index + 1; newBeginIndex < _endValidElementsIndex; ++newBeginIndex) {
+                            const NamedData<DataType>& data = _vNamedDatas[newBeginIndex];
+                            if (!data.IsValid()) {
+                                continue;
+                            }
+                            _beginValidElementsIndex = newBeginIndex;
+                            break;
+                        }
+                    }
+                }
+                else if (index + 1 == _endValidElementsIndex) {
+                    if (_numValidElements == 1) {
+                        _endValidElementsIndex = _beginValidElementsIndex + 1;
+                    }
+                    else {
+                        for (size_t newEndIndex = _endValidElementsIndex - 1; newEndIndex > _beginValidElementsIndex; --newEndIndex) {
+                            const NamedData<DataType>& data = _vNamedDatas[newEndIndex - 1];
+                            if (!data.IsValid()) {
+                                continue;
+                            }
+                            _endValidElementsIndex = newEndIndex;
+                            break;
+                        }
+                    }
+                }
+            }
+            OPENRAVE_ASSERT_OP(_numValidElements, <= ,(_endValidElementsIndex - _beginValidElementsIndex));
+            return;
+        }
+        OPENRAVE_ASSERT_FORMAT(false, "nameId=%d is not found", nameId, OpenRAVE::ORE_InvalidArguments);
+    }
+
+    /// \brief finds element for key
+    /// \param nameId key to find value for
+    /// \param value value for key.
+    /// \return whether key is found.
+    bool Find(uint64_t nameId, DataType& value) const
+    {
+        OPENRAVE_ASSERT_FORMAT0(nameId != 0, "nameId cannot be 0. 0 is reserved for invalid", OpenRAVE::ORE_InvalidArguments);
+
+        for (const NamedData<DataType>& customData : _vNamedDatas) {
+            if (customData.nameId == nameId) {
+                value = customData.data;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /// \brief inserts element
+    /// \param nameId key to insert value for
+    /// \param data value for key.
+    void Insert(uint64_t nameId, const DataType& data)
+    {
+        OPENRAVE_ASSERT_FORMAT0(nameId != 0, "nameId cannot be 0. 0 is reserved for invalid", OpenRAVE::ORE_InvalidArguments);
+
+        int64_t invalidIndex = -1;
+        for (size_t index = 0; index < _vNamedDatas.size(); ++index) {
```

Comment:
you should make a comment here that you are looking through the entire array in order to find invalid `nameId`s too

Reply:
Thanks added.